runtime/gc_glue_java/ScavengerDelegate.cpp

```diff
@@ -648,6 +656,104 @@ MM_ScavengerDelegate::private_shouldPercolateGarbageCollect_classUnloading(MM_En
 	return shouldGCPercolate;
 }
 
+void
+MM_ScavengerDelegate::private_SortAllHotField()
+{	
+	/* update hottest fields for all elements of the hotFieldClassInfoPool where isClassHotFieldListDirty is true */
+	if (_javaVM->hotFieldClassInfoPool != NULL && ((_extensions->scavengerStats._gcCount  % _extensions->gcCountBetweenHotFieldSort) == 0)) {
+		pool_state hotFieldClassInfoPoolState;
+		omrthread_monitor_enter(_javaVM->hotFieldClassInfoPoolMutex);
+		J9ClassHotFieldsInfo *hotFieldClassInfoTemp = (J9ClassHotFieldsInfo *)pool_startDo(_javaVM->hotFieldClassInfoPool, &hotFieldClassInfoPoolState);
+		
+		/* sort hot field list for the class if the hot field list of the class is dirty */
+		while (NULL != hotFieldClassInfoTemp && hotFieldClassInfoTemp->consecutiveHotFieldSelections != U_8_MAX) {
+			if (hotFieldClassInfoTemp->isClassHotFieldListDirty) {
+				private_SortClassHotFieldList(hotFieldClassInfoTemp);
+			}
+			hotFieldClassInfoTemp = (struct J9ClassHotFieldsInfo*)pool_nextDo(&hotFieldClassInfoPoolState);
+		}
+		omrthread_monitor_exit(_javaVM->hotFieldClassInfoPoolMutex);
+	}
+	/* If adaptiveGcCountBetweenHotFieldSort, update the gc count required between sorting all hot fields as the application runs longer */
+	if (_extensions->adaptiveGcCountBetweenHotFieldSort && _extensions->gcCountBetweenHotFieldSort < _extensions->gcCountBetweenHotFieldSortMax && _extensions->scavengerStats._gcCount % INCREMENT_GC_COUNT_BETWEEN_HOT_FIELD_SORT == 0) {
+		_extensions->gcCountBetweenHotFieldSort++;
+	}
+	/* If hotFieldResettingEnabled, update the gc count required between resetting all hot fields */
+	if (_extensions->hotFieldResettingEnabled && _extensions->scavengerStats._gcCount % _extensions->gcCountBetweenHotFieldReset == 0) {
+		private_ResetAllHotFieldData();
+	}
+}
+
+void
+MM_ScavengerDelegate::private_SortClassHotFieldList(J9ClassHotFieldsInfo* hotFieldClassInfo) {
+	/* store initial hot field offsets before hotFieldClassInfo hot field offsets are updated */
+	U_8 initialHotFieldOffset1 = hotFieldClassInfo->hotFieldOffset1;
+	U_8 initialHotFieldOffset2 = hotFieldClassInfo->hotFieldOffset2;
+
+	/* compute and update the hot fields for each class */
+	if (hotFieldClassInfo->hotFieldListLength == 1) {
```

Comment:
constant first

Reply:
Is this special case needed for correctness, or this is just a speed optimization? I would not keep just for speed.